File: src/day02.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct cubes
{
	int red;
	int blue;
	int green;
};

int ends_width(const char *str, const char *suffix)
{
	const size_t str_len = strlen(str);
	const size_t suffix_len = strlen(suffix);

	if (strncmp(str + str_len - suffix_len, suffix, suffix_len) == 0)
	{
		return 1;
	}

	if (strncmp(str + str_len - suffix_len - 1, suffix, suffix_len) == 0)
	{
		return 1;
	}

	return 0;
}

int max(const int val1, const int val2)
{
	return val1 > val2 ? val1 : val2;
}

struct cubes get_max_count(char *str)
{
	struct cubes result = {};

	const char *previous = NULL;
	const char *current = strtok(str, " ");

	while (current != NULL)
	{
		if (ends_width(current, "red"))
		{
			result.red = max(result.red, atoi(previous));
		}
		else if (ends_width(current, "green"))
		{
			result.green = max(result.green, atoi(previous));
		}
		else if (ends_width(current, "blue"))
		{
			result.blue = max(result.blue, atoi(previous));
		}

		previous = current;
		current = strtok(NULL, " ");
	}

	return result;
}

int possible(const struct cubes *game)
{
	return game->red <= 12 && game->green <= 13 && game->blue <= 14;
}

int main()
{
	FILE *file = fopen("input/day02", "r");

	size_t index = 0;
	int sum1 = 0;
	int sum2 = 0;
	char buffer[256];

	while (fgets(buffer, sizeof(buffer), file))
	{
		index++;
		const struct cubes game = get_max_count(buffer);
		if (possible(&game))
		{
			sum1 += index;
		}

		sum2 += game.red * game.green * game.blue;
	}

	fclose(file);

	printf("[01] sum: %i\n", sum1);
	printf("[02] sum: %i\n", sum2);
	return 0;
}

File: day02.lua
---@class Cubes
---@field red number
---@field green number
---@field blue number

---@param str string
---@param suffix string
---@return boolean
function ends_width(str, suffix)
	return str:sub(#str - #suffix, #suffix) == suffix
		or str:sub(#str - #suffix - 1, #suffix) == suffix
end

---@param str string
---@return Cubes
function get_max_count(str)
	---@type Cubes
	local result = {
		red = 0,
		green = 0,
		blue = 0,
	}

	local matches = string.gmatch(str, "%S+")
	local i = 0
	local previous
	for current in matches do
		i = i + 1
		if ends_width(current, "red") then
			result.red = math.max(result.red, tonumber(previous))
		elseif ends_width(current, "green") then
			result.green = math.max(result.green, tonumber(previous))
		elseif ends_width(current, "blue") then
			result.blue = math.max(result.blue, tonumber(previous))
		end
		previous = current
	end

	return result
end

---@param game Cubes
---@return boolean
function is_possible(game)
	return game.red <= 12 and game.green <= 13 and game.blue <= 14
end

local sum1 = 0
local sum2 = 0
local index = 0

for line in io.lines("input/day02") do
	index = index + 1
	local game = get_max_count(line)
	if is_possible(game) then
		sum1 = sum1 + index
	end
	sum2 = sum2 + game.red * game.green * game.blue
end

print("[01] sum:", sum1)
print("[02] sum:", sum2)